src/sources/kafka.rs

```diff
@@ -237,34 +242,59 @@ async fn kafka_source(
     acknowledgements: bool,
 ) -> Result<(), ()> {
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, shutdown.clone());
-    let finalizer = finalizer.map(Arc::new);
-    if let Some(finalizer) = &finalizer {
+    let (mut finalizer, mut ack_stream) =
+        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, None);
+
+    let (tx_callbacks, mut rx_callbacks) = mpsc::unbounded_channel();
```

Comment:
Assuming there isn't a deadlock concern, it could probably even be lower, like 1 or 2. Rebalance events should be _rare_ compared to the flow of events.

Reply:
Yeah, they should be rare, but I think an event comes in per partition revoked, so ideally we'd want a high enough bound for the number of partitions an individual consumer serves on average to try and avoid having any of the `librdkafka` threads block on capacity.

I'm certainly not a by-day Kafka wrangler, but from what I know about common Kafka sizing exercises, we'd probably expect that number to be somewhere in the 10-20 range.... and the events (enum) are gonna be a single machine word, so 2 vs 128 vs 16 is gonna be indistinguishable noise in memory usage.

Reply:
For each invocation of the `post_rebalance` method on the client Context, we get a `TopicPartitionList` that has all of the partitions being assigned or revoked, so I think this will typically be called 2 times during each rebalance event -- once with a `Revoke` and once with an `Assign` (it's still called even if the list is empty). It shouldn't scale with the number of partitions since the `Revoke` events contain a list of the affected partitions.

There is a potential deadlock here though: with the current design `post_rebalance` runs as a result of reading messages off of the stream, when the client gets a "time to rebalance" notification instead of new messages from the broker. So if the `post_rebalance` method blocks, it will block that entire main `select!` that both `stream.next()` and `rx_callbacks.recv()` are branches of, and the channel will never be read.

All this to say..I think a bound of 1 would work fine here since we're only sending for revokes,  _as long as_ the receiver is processed before another rebalance event. However, since rebalances, while typically rare, can happen in clusters in certain circumstances, I would think a slightly higher bound like 16 is probably sufficient.

Reply:
As noted in the [conversation below](https://github.com/vectordotdev/vector/pull/14846/files#r1008317713), leaving this as an unbounded channel for now due to the "sync `fn` within async task" context of the sender, which makes using a bounded channel difficult.